File: java/src/test/java/mrk/hackerrank/HackerRank04Test.java
package mrk.hackerrank;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class HackerRank04Test {

    private static Stream<Arguments> testData() {
        return Stream.of(
                Arguments.of(new int[]{5, 6, 7}, new int[]{3, 6, 10}, new int[]{1, 1}),
                Arguments.of(new int[]{1, 2, 3}, new int[]{3, 2, 1}, new int[]{1, 1}),
                Arguments.of(new int[]{1, 1, 1}, new int[]{1, 1, 1}, new int[]{0, 0}),
                Arguments.of(new int[]{3, 3, 3}, new int[]{1, 1, 1}, new int[]{3, 0}),
                Arguments.of(new int[]{1, 1, 1}, new int[]{3, 3, 3}, new int[]{0, 3})
        );
    }

    @ParameterizedTest
    @MethodSource("testData")
    void test(int[] a, int[] b, int[] expected) {
        Assertions.assertArrayEquals(expected, new HackerRank04().compareTriplets(a, b));
    }

    @ParameterizedTest
    @MethodSource("invalidInputs")
    void testInvalidInput(int[] a, int[] b) {
        Assertions.assertThrows(IllegalArgumentException.class, () -> new HackerRank04().compareTriplets(a, b));
    }

    private static Stream<Arguments> invalidInputs() {
        return Stream.of(
                Arguments.of(new int[]{1, 2}, new int[]{1, 2, 3}),
                Arguments.of(new int[]{1, 2, 3}, new int[]{1, 2}),
                Arguments.of(new int[]{}, new int[]{1, 2, 3}),
                Arguments.of(new int[]{1, 2, 3, 4}, new int[]{1, 2, 3})
        );
    }
}

File: kotlin/src/test/kotlin/HackerRank04Test.kt
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class HackerRank04Test {
    @ParameterizedTest(name = "{0}")
    @MethodSource("hackerRank04Data")
    fun test_hackerRank04(
        name: String,
        a: Array<Int>,
        b: Array<Int>,
        output: Array<Int>,
    ) {
        assertArrayEquals(output, hackerRank04(a, b))
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("invalidInputs")
    fun test_hackerRank04_invalidInput(
        name: String,
        a: Array<Int>,
        b: Array<Int>,
    ) {
        assertThrows(IllegalArgumentException::class.java) { hackerRank04(a, b) }
    }

    companion object {
        @JvmStatic
        fun hackerRank04Data() =
            listOf(
                Arguments.of("a wins 1, b wins 1", arrayOf(5, 6, 7), arrayOf(3, 6, 10), arrayOf(1, 1)),
                Arguments.of("a wins 1, b wins 1 reversed", arrayOf(1, 2, 3), arrayOf(3, 2, 1), arrayOf(1, 1)),
                Arguments.of("draw", arrayOf(1, 1, 1), arrayOf(1, 1, 1), arrayOf(0, 0)),
                Arguments.of("a wins all", arrayOf(3, 3, 3), arrayOf(1, 1, 1), arrayOf(3, 0)),
                Arguments.of("b wins all", arrayOf(1, 1, 1), arrayOf(3, 3, 3), arrayOf(0, 3)),
            )

        @JvmStatic
        fun invalidInputs() =
            listOf(
                Arguments.of("a too short", arrayOf(1, 2), arrayOf(1, 2, 3)),
                Arguments.of("b too short", arrayOf(1, 2, 3), arrayOf(1, 2)),
                Arguments.of("a empty", emptyArray<Int>(), arrayOf(1, 2, 3)),
                Arguments.of("a too long", arrayOf(1, 2, 3, 4), arrayOf(1, 2, 3)),
            )
    }
}
